3D Tiles - Dynamic Screen Space Error

lilleyse opened the pull request:
For #3241 

Like terrain fog, this enables a dynamic screen space error for rendering lower detail tiles based on distance. I mainly tested this with replacement refinement tilesets and I see about 20% reduction in the number of visited tiles for street-level views for city-like tilesets and anywhere from 50% to 70% improvement for more terrain-like datasets without too much loss in visual quality. These percentages fluctuate depending on the `maximumScreenSpaceError` that is applied, and is more favorable for lower values. 

The settings can be tweaked by changing `tileset.usesDynamicScreenSpaceError`, `tileset.dynamicScreenSpaceErrorDensity`, and `tileset.dynamicScreenSpaceErrorFactor`.

lilleyse commented on Source/Scene/Fog.js:
```diff
@@ -131,7 +131,7 @@ define([
         // Fade fog in as the camera tilts toward the horizon.
         var positionNormal = Cartesian3.normalize(camera.positionWC, scratchPositionNormal);
-        var dot = CesiumMath.clamp(Cartesian3.dot(camera.directionWC, positionNormal), 0.0, 1.0);
+        var dot = Math.abs(Cartesian3.dot(camera.directionWC, positionNormal));
```
I changed this because I noticed that any views that are looking from horizon-view to birds-eye-view would result in a negative dot product. So all views between horizon and birds-eye are considered to be in complete fog, but really it should adjust.

lilleyse replied:
By the way, this does not really affect anything in practice, so if anything we could just remove this code.

pjcozzi replied:
Up to you and @bagnell

bagnell replied:
You probably want to revert this. You can look up at terrain (try at Everest, it's pretty easy) in which case we want to clamp at zero.

bagnell replied:
Actually, after talking offline. This will be fine. @lilleyse is going to update the tests.

pjcozzi commented:
@bagnell can you also take a quick look at this?

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -70,6 +72,7 @@ define([
      * @param {Matrix4} [options.modelMatrix=Matrix4.IDENTITY] A 4x4 transformation matrix that transforms the tileset's root tile.
      * @param {Number} [options.maximumScreenSpaceError=16] The maximum screen-space error used to drive level-of-detail refinement.
      * @param {Boolean} [options.refineToVisible=false] Whether replacement refinement should refine when all visible children are ready. An experimental optimization.
+     * @param {Boolean} [options.useDynamicScreenSpaceError=false] Reduce the screen space error for tiles that are further away from the camera.
```
We generally don't start variables with `use`, e.g., above is just `refineToVisible`, not `useRefineToVisible`, so this could just be `dynamicScreenSpaceError` or `scaleScreenSpaceErrorByDistance`.

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -131,6 +134,31 @@ define([
+         *
+         * @see Fog
+         */
+        this.useDynamicScreenSpaceError = defaultValue(options.useDynamicScreenSpaceError, false);
```
Have you tried this with a few different datasets?  Perhaps it can default to terrain like terrain fog.

lilleyse replied:
I think defaulting to false is safer for now. I've tried on a few datasets and having it on isn't too different but technically it does break the concept of a constant screen space error and we don't actually visualize it with fog.

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -131,6 +134,31 @@ define([
+         *
+         * @see Fog.density
+         */
+        this.dynamicScreenSpaceErrorDensity = 0.00278;
```
Here and below, what is the range and how does the scale work in general?  I don't think there is enough doc here for a user to have a intuitive sense of how to set this.

Can we also expose this parameters in the Sandcastle example...later they would be in the 3D Tiles inspector?

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -788,19 +816,57 @@ define([
+        var positionCartographic = camera.positionCartographic;
+        var height = positionCartographic.height;
+
+        // The range where the density starts to lessen
```
How did you come up with these and should they be based on the tileset's bounding box?

lilleyse replied:
I added a part-way solution. It now uses the bounding volume's center as a starting point. However I wasn't sure how to get the "height" of the tileset in a consistent way. It's easy for regions but obb's would need to be aligned z-up, and for bounding spheres it's hard to know at all.

pjcozzi replied:
If the max height is needed (not sure why, we should be thinking in general 3D for 3D Tiles), I think it can be computed pretty easily for spheres.  OBBs could be harder, but if they are too hard for now; we could approximate them with a sphere for this computation.

For spheres, project the point onto the surface using the geodetic surface normal.  Subtract that distance from the radius, then go from the surface point along the geodetic normal to (radius - that

pjcozzi replied:
...(radius - that subtracted distance).  Would need to think more deeply if this works for all ellipsoids or just spheres.

lilleyse replied:
Yeah that way could work for most cases, we may need to query the terrain height in order to get accurate results for tilesets that are at higher elevations.

The max height is used as an approximate way to see how far the camera is from the tileset. When the camera is high enough I don't apply any dynamic sse because it tends to make the whole tileset pretty blurry. My calculations are really meant to apply dynamic SSE at street level views where a wide range of LODs are visible at the same time. At the moment height seemed to be the best way to measure this.

pjcozzi commented:
Just those comments.  Will also test this with some datasets once the Sandcastle example has more options.

pjcozzi commented:
When fog is used with a dynamic SSE for terrain, we also stop requesting tiles that are completely in fog.

It is worth doing something similar here with a max distance?

lilleyse commented:
> When fog is used with a dynamic SSE for terrain, we also stop requesting tiles that are completely in fog. It is worth doing something similar here with a max distance?

We could do that, but also the tiles in the distance should be pretty low res to begin with so there won't be that many requests saved.

lilleyse commented:
Though it does depend on the

pjcozzi commented:
> We could do that, but also the tiles in the distance should be pretty low res to begin with so there won't be that many requests saved.

This widely depends on the tileset.  Consider a root tile with 300 children, or just a uniform grid.

pjcozzi commented:
Any concerns if I merge this into `3d-tiles-temp` for testing?

lilleyse commented:
> Any concerns if I merge this into 3d-tiles-temp for testing?

Nope, go ahead.

pjcozzi commented:
My testing and the code look good.  I still don't think the doc for `dynamicScreenSpaceErrorDensity` and `dynamicScreenSpaceErrorFactor` make it obvious how to use them, e.g., explain-ish how they are non-linear.  Perhaps the terrain doc also should be improved.

I'll also hold off on merging until the local coordinate system change we discussed offline.

lilleyse commented:
Updated.

> When fog is used with a dynamic SSE for terrain, we also stop requesting tiles that are completely in fog.

> It is worth doing something similar here with a max distance?

I could be wrong, but I think it's just doing the same thing as here, requesting less because it refines less. But the uniform grid type scenario for 3d tiles is definitely worth handling. Is it ok if I handle this in another PR? I also want to clean up some of the checks we do (check sse, check visibility, compute distance, sort by distance, check request volume) that we do in multiple places.

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -131,6 +142,42 @@ define([
+         *
+         * @see Fog.density
+         */
+        this.dynamicScreenSpaceErrorDensity = 0.00278;
```
Any reason this and `dynamicScreenSpaceErrorFactor` should not also be constructor arguments?

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -788,19 +835,107 @@ define([
+        var root = tileset._root;
+        var tileBoundingVolume = root.contentBoundingVolume;
+
+        if (tileBoundingVolume instanceof TileBoundingRegion) {
```
Do you know how fast `instanceof` is?  If it isn't basically free we should not use it here.

pjcozzi replied:
Ah, this is only for the `root` tile.  Probably not an issue, but we should still research.

pjcozzi replied:
That is actual kind of slow I bet compared to, for example, `===` with numbers.  Regardless, it is fine here, of course.

lilleyse replied:
I used this test: https://jsperf.com/instanceof-performance/2.

The code is not much of a hotspot, it is called once per frame.

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -788,19 +835,107 @@ define([
+            }
+        }
+
+        // The range where the density starts to lessen. Start at the quarter height of the tileset.
```
Shouldn't this be a user option?

pjcozzi commented on Source/Scene/Cesium3DTileset.js:
```diff
@@ -788,19 +835,107 @@ define([
+    var scratchPosition = new Cartesian3();
+    var scratchDirection = new Cartesian3();
+
+    function updateDynamicScreenSpaceError(tileset, frameState) {
```
How is test coverage?  Does the new test cover all the paths in this function?

lilleyse commented:
I'll look at this

lilleyse commented:
Updated. I addressed the remaining points. I just have to add one more bounding volume test, and will update this issue when its ready (should be later today).

pjcozzi commented:
CC @pmconne

pjcozzi commented:
Looks good, let me know when the tests are finished.

lilleyse commented:
The last test is done now.

pjcozzi commented:
Do these tests fail for you?

```
Scene/GlobeSurfaceTileProvider fog culls tiles in full fog
Expected [ 255, 0, 0, 255 ] to equal [ 0, 0, 0, 255 ].
Error: Expected [ 255, 0, 0, 255 ] to equal [ 0, 0, 0, 255 ].
    at stack (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1475:17)
    at buildExpectationResult (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1445:14)
    at Spec.expectationResultFactory (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:584:18)
    at Spec.addExpectationResult (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:332:34)
    at Expectation.addExpectationResult (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:528:21)
    at Expectation.toEqual (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1399:12)
    at http://localhost:8080/Specs/Scene/GlobeSurfaceTileProviderSpec.js:391:48
    at Promise.then (http://localhost:8080/Source/ThirdParty/when.js:196:34)
    at http://localhost:8080/Source/ThirdParty/when.js:297:13
    at processQueue (http://localhost:8080/Source/ThirdParty/when.js:647:4)
Scene/GlobeSurfaceTileProvider fog culls tiles because of increased SSE
Expected [ 255, 0, 0, 255 ] not to equal [ 255, 0, 0, 255 ].
Error: Expected [ 255, 0, 0, 255 ] not to equal [ 255, 0, 0, 255 ].
    at stack (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1475:17)
    at buildExpectationResult (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1445:14)
    at Spec.expectationResultFactory (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:584:18)
    at Spec.addExpectationResult (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:332:34)
    at Expectation.addExpectationResult (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:528:21)
    at Expectation.toEqual (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1399:12)
    at http://localhost:8080/Specs/Scene/GlobeSurfaceTileProviderSpec.js:417:52
    at Promise.then (http://localhost:8080/Source/ThirdParty/when.js:196:34)
    at http://localhost:8080/Source/ThirdParty/when.js:297:13
    at processQueue (http://localhost:8080/Source/ThirdParty/when.js:647:4)
```

lilleyse commented:
Ah those are caused by this change: https://github.com/AnalyticalGraphicsInc/cesium/pull/4307#r79319602

Because of the old math, fog would be applied in bird's eye view, which is the view these tests use. But my change prevents fog from happening for those views. @bagnell can you check that my code change makes sense?

pjcozzi commented:
@bagnell thoughts here? We want to merge this soon.

lilleyse commented:
Updated. Those tests pass now.

pjcozzi commented:
👍
